### ISSUE TITLE
Screen capture : createLocalVideoTrack creating webcam track instead of screen

### ISSUE THREAD

**Issue by theDude30** [twilio/twilio-video.js#99]
Hi,
I'm trying to do screen capture, I've managed to do the following:
1. get token.
2. install extension.
3. select the desktop and get sourceID response.
4. used navigator.getUserMedia with the given sourceID.

Now when i do "Video.createLocalVideoTrack with the stream.getVideoTracks()[0]" it start capturing the webcam.
What am i missing? 

 - [X ] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [X ] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [X ] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.

**Code to reproduce the issue:**

   navigator.mediaDevices.getUserMedia({
                                    video: {
                                        mandatory: {
                                            chromeMediaSource: 'desktop',
                                            chromeMediaSourceId: respData.sourceId
                                        }
                                    }
                                }).then(function (stream) {
                                   var localStream= Video.createLocalVideoTrack(stream.getVideoTracks()[0]).then(function(localTracks) {
                                        var currentLocalTrack=localTracks;
                                        var localMediaContainer = document.getElementById('localVideo');
                                            localMediaContainer.appendChild(currentLocalTrack.attach());
                                            Twilio.Video.connect(data.token, {
                                                name: room,
                                                tracks: [currentLocalTrack],
                                                audio: false,
                                                video:false
                                            }).then(function (twRoom) {
                                                twRoom.localParticipant.addTrack(currentLocalTrack);
                                                console.log('Connected to room ' + twRoom.name + ' with screen sharing');
                                        });
                                    });
                                });

**Expected behavior:**

The screen should be shared(i even got the google message that the screen is being shared)

**Actual behavior:**

It captures my webcam.

**Software versions:**

 - [X ] Browser(s):Chrome 58.0.3029.96 (64-bit)
 - [X ] Operating System: OSX 10.12.4
 - [ X] twilio-video.js: 1.0.0
 - [ X] Third-party libraries (e.g., Angular, React, etc.): None(basic javascript+jquery).

**Comment by manjeshbhargav** [comment]
Hi @theDude30 ,

`createLocalVideoTrack` accepts a video `MediaTrackConstraints` object and not a `MediaStreamTrack`. Alternatively, this will work:
```javascript
const { connect, createLocalVideoTrack } = require('twilio-video');

(async () => {
  const desktopTrack = await createLocalVideoTrack({
    mandatory: {
      chromeMediaSource: 'desktop',
      chromeMediaSourceId: 'media-source-id-from-extension'
    }
  });

  const room = await connect('my-token', {
    name: 'my-room-name',
    tracks: [desktopTrack]
  });
})();
```

Hope this unblocks you. Please let me know if you have any more questions.

- Manjesh

**Comment by theDude30** [comment]
var localStream= Video.createLocalVideoTrack({
                                       video: {
                                           mandatory: {
                                               chromeMediaSource: 'desktop',
                                               chromeMediaSourceId: respData.sourceId
                                           }
                                       }
                                   }).then(function(localTracks) {
                                        var currentLocalTrack=localTracks;
                                        var localMediaContainer = document.getElementById('localVideo');
                                            localMediaContainer.appendChild(currentLocalTrack.attach());
                                            Twilio.Video.connect(data.token, {
                                                name: room,
                                                tracks: [localTracks],
                                                audio: false,
                                                video:false
                                            }).then(function (twRoom) {
                                                twRoom.localParticipant.addTrack(currentLocalTrack);
                                                console.log('Connected to room ' + twRoom.name + ' with screen sharing');
                                        });
                                    });

**Comment by manjeshbhargav** [comment]
Hi @theDude30 ,

Try this and let me know if it works:

```javascript
Video.createLocalVideoTrack({
  mandatory: {
    chromeMediaSource: 'desktop',
    chromeMediaSourceId: respData.sourceId
  }
}).then(function(desktopTrack) {
  var localMediaContainer = document.getElementById('localVideo');
  localMediaContainer.appendChild(desktopTrack.attach());
  return Video.connect(data.token, { name: room, tracks: [ desktopTrack ] });
}).then(function(room) {
  console.log('Connected to room ' + room.name + ' with screen sharing!');
});
```

Manjesh Malavalli

**Comment by theDude30** [comment]
Works great thanks

**Comment by manjeshbhargav** [comment]
@theDude30 ,

Cool! I'll close the issue then.

Manjesh Malavalli

**Comment by Arthur447** [comment]
@manjeshbhargav,

How to do for this scenario : 
1) Twlio.connect with audio and video
2) User click "Share screen" after
I would like to get Screen (navigator.getUserMedia) and then attach it to the room tracks,

Is it possible ?

Thanx

**Comment by manjeshbhargav** [comment]
Hi @Arthur447 ,

We have a screen sharing guide for Chrome that you can access [here](https://www.twilio.com/docs/api/video/screen-capture-chrome).

Your specific scenario is possible. Here's how you can do it:
```javascript
Video.connect('my-token', { name: 'my-room' }).then(function(room {
  $('#sharescreen').click(function() {
    Video.createLocalVideoTrack({
      chromeMediaSource: 'desktop',
      chromeMediaSourceId: 'media-source-id-from-chrome-screenshare-extension'
    }).then(function(track) {
      room.localParticipant.publishTrack(track);
    });
  });
});
```

**Comment by Arthur447** [comment]
Hi @manjeshbhargav 

I managed with room.localParticipant.addTracks(screen_tracks) and it works well, thanx ;)

however, i have problem with Firefoxv56 -> Firefoxv56, the demo (https://github.com/twilio/video-quickstart-js) is working with Chrome->Chrome and Chrom<->Firefox but when we use Firefox->Firefox, there is a warning : "ICE failed, add a STUN server and see about:webrtc for more details"

When i look in about:webrtc i notice than Candidat local and Candidat distant use both local IP so they can't find each other. 

Do you wan't me to create a new issue for that ? Do you have any suggestion ?

Best,

**Comment by manjeshbhargav** [comment]
Hi @Arthur447 ,

Are you running the quickstart app on Firefox, or one of the examples in the quickstart project? If so can you let me know which example you are running?

Thanks,

Manjesh Malavalli

**Comment by Arthur447** [comment]
Hi @manjeshbhargav ,

I am running the quickstart app on Firefox. I just use twilio CDN library, as you can see here : view-source:https://www.communautel.fr/twilio line 389.

Arthur

**Comment by manjeshbhargav** [comment]
Hi @Arthur447 ,

I tried to connect to a Room in your app, but I'm getting an error saying "Participant identity is invalid". How are you generating the Room token?

Thanks,

Manjesh Malavalli

**Comment by Arthur447** [comment]
Hi @manjeshbhargav ,

I am generating token from server side with Twilio PHP Library v4 but only for authenticated users, your error is normal. 

I think that i found the issue and solved it yesterday with this code

```
                    var connectOptions = {
                        name: roomName,
                        logLevel: 'debug',
                        iceServers: [ { 'urls': 'stun:stun.l.google.com:19302'} ] 
                    };
```

As you can see, i use Google stun server and it worked with my local tests.
I will do more tests in real conditions and keep you inform if you want,

Thanks,
Arthur

**Comment by manjeshbhargav** [comment]
@Arthur447 ,

Thanks for the update. Sure, I'd love more information about the ICE issue that you're facing. If possible, can you run your app on Firefox without the iceServers and send me the debug logs from the console?

Thanks,

Manjesh Malavalli

**Comment by Arthur447** [comment]
Hi @manjeshbhargav 

I have done 4 tests : 
Tests 1, 2 and 3 with Google STUN serve : all tests are OK
Test 4 : default Twilio Iceserver with Firefox=>Firefox : test Failed

You can see logs in attached file
[tests_results.txt](https://github.com/twilio/twilio-video.js/files/1437004/tests_results.txt)

Please, let me know if you see something,

Best

**Comment by shrutigv** [comment]
@manjeshbhargav,

I am trying to create two local streams with webcam and external camera plugged in. using createLocalVideoTrack and then attach tracks but unable to stream could you please let me know any other api to serve my usecase?

**Comment by manjeshbhargav** [comment]
Hi @shrutigv ,

Thanks for writing in with your question. Can you open a new issue and explain a bit more in detail the use-case that you are looking for, and what exactly is the problem you are running into?

Thanks,

Manjesh Malavalli
JSDK Team
